File: 128.longest-consecutive-sequence.py
#
# @lc app=leetcode id=128 lang=python3
#
# [128] Longest Consecutive Sequence
#

# Time:  O()
# Space: O()

from typing import List


# @lc code=start
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        res, num2len = 0, {num: 0 for num in nums}
        for num in nums:
            if num2len[num] == 0:
                num2len[num] += 1
                left = num2len.get(num - 1, 0)
                right = num2len.get(num + 1, 0)
                num_len = left + 1 + right
                num2len[num - left] = num_len
                num2len[num + right] = num_len
                res = max(res, num_len)
        return res


# @lc code=end

File: 128.longest-consecutive-sequence.js
/*
 * @lc app=leetcode id=128 lang=javascript
 *
 * [128] Longest Consecutive Sequence
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function (nums) {
  let res = 0;
  const lookup = new Set(nums);
  for (const num of nums) {
    if (!lookup.has(num)) {
      continue;
    }
    lookup.delete(num);
    let prev = num - 1;
    let next = num + 1;
    while (lookup.has(prev)) {
      lookup.delete(prev--);
    }
    while (lookup.has(next)) {
      lookup.delete(next++);
    }
    res = Math.max(res, next - prev - 1);
  }
  return res;
};
// @lc code=end
